SARS-CoV-2-IHMV - VERSO
=================

In this repository we provide: 

1. A set of scripts to perform *variant calling* from Illumina
SARS-CoV-2 raw sequencing data.
2. R scripts to reproduce Step #1 of the VERSO pipeline on Dataset #1
of the original article.
3. A Jupyter notebook to reproduce Step #2 of the VERSO pipeline on
Dataset #1 of the original article.

Please feel free to contact us if you have problems running our
scripts at daniele.ramazzotti1@gmail.com or
d.maspero@campus.unimib.it.


### Reference

*Quantification of intra-host genomic diversity of
SARS-CoV-2 allows a high-resolution characterization of viral
evolution and reveals functionally convergent variants*, Daniele
Ramazzotti, Fabrizio Angaroni, Davide Maspero, Carlo
Gambacorti-Passerini, Marco Antoniotti, Alex Graudenzi, Rocco
Piazza. Preprint available at
https://www.biorxiv.org/content/10.1101/2020.04.22.044404v2


### Requirements

The current *VERSO* setup assumes a UN\*X/Linux platform, altough
*VERSO*, per se, is written in R and should therefore run everywhere.

The variant calling scripts use the UN\*X shell (`bash`, although no
`bash`-only features have been used).


### License

Please check the LICENSE file in the top directory for licensing
information.
